### ISSUE TITLE
[casual_completion] introduce component to perform causal completion on the DAG

### ISSUE THREAD

**Issue by akichidis** [MystenLabs/narwhal#184]
### Story

Parent story: #

### Description

Will add more context soon...

**Comment by huitseeker** [comment]
Yes, the issue accurately captures the story.
- note the synchronization is very similar to the one present in the `header_waiter` already:
https://github.com/MystenLabs/narwhal/blob/7566e3e94bba636092bce523fe5d0da27facf8d8/primary/src/header_waiter.rs#L190-L235
- Simply hooking into the `header_waiter` to perform causal completion would work, except we do not have this distinguished node from which to extract the certificate, and would want to jump directly to the lucky broadcast phase,
- the code pointer you have to the core is highly relevant, because it indicates what happens in the process of delivering a message: check signatures on the current payload, then fetch parents, recursively deliver them completely, then proceed with delivering the current header,
- we should make sure checking signatures is the first step (in application of the [doom principle](https://moxie.org/2011/12/13/the-cryptographic-doom-principle.html)),
- The difference with the background synchronization occurring in the core is that in the context of read_causal, we want to block (and possibly be more aggressive in broadcasts), because we're latency-sensitive, whereas the core implementation is asynchronous everywhere (and throughput-sensitive).

**Comment by akichidis** [comment]
Given the latest discussions with @huitseeker the following problem(s) have been arisen which need to be tackled to ensure that we don't break system's assumptions and we remain consistent in our approach:

### 1. Storing the fetched `certificates` & `headers` from the causal completion

As it has been described on this issue, to perform the causal completion during the `read_causal` request _in a blocking manner_ we'll use the `block_synchronizer` to fetch any missing certificates as we traverse the DAG. By this process a list of "missing" certificates will be formed which we managed to get from other primary peers. Next step would be to:
* process and store the `certificate`
* process and store the `header`

#### A. Processing the certificate
To successfully process the certificate and end up storing in the `certificate_store`, we have to follow the existing approach and basically do the following:

1. [verify the certificate](https://github.com/MystenLabs/narwhal/blob/06203d926ec214768ffd83477e3606bb4d4855b0/primary/src/core.rs#L367) ✅ (we already do as part of the `blocksynchronizer`)
2. ensure [all parent certificates exist in certificate_store](https://github.com/MystenLabs/narwhal/blob/main/primary/src/core.rs#L288-L301) ❌ (we don't currently do in `block_synchronizer`)
3. trigger [block proposer in case enough have been gathered](https://github.com/MystenLabs/narwhal/blob/main/primary/src/core.rs#L304-L315) ❌ (we don't currently do in `block_synchronizer`)

#### B. Processing the header
To successfully process a header and store in the `header_store`, the following need to happen:
1. [verify the header](https://github.com/MystenLabs/narwhal/blob/06203d926ec214768ffd83477e3606bb4d4855b0/primary/src/core.rs#L335) ❌ (we don't currently do in block_synchronizer)
2. [ensure payload (batches) is available](https://github.com/MystenLabs/narwhal/blob/06203d926ec214768ffd83477e3606bb4d4855b0/primary/src/core.rs#L164) ❌ (we don't currently do in block_synchronizer)
3. ensure [parents are available](https://github.com/MystenLabs/narwhal/blob/06203d926ec214768ffd83477e3606bb4d4855b0/primary/src/core.rs#L173) ❌ (we don't currently do in block_synchronizer)
4. [verify the stake (should be 2f+1)](https://github.com/MystenLabs/narwhal/blob/06203d926ec214768ffd83477e3606bb4d4855b0/primary/src/core.rs#L181-L191) ❌ (we don't currently do in block_synchronizer)

We might not really need to perform some of the above. For example the A.3, do we really need it or do we want to run on assumptions that our node will eventually process some of those certificates anyways?

The one thing that we are certain at this point, given the current implementation and from the protocol, is that **we can store the fetched certificates without needing to already have the payload fetched or the header already processed**.


### 2. `block_synchronizer` and `header_waiter` can lead to duplicate processing

When a `certificate` or a `header` is getting processed via [core](https://github.com/MystenLabs/narwhal/blob/main/primary/src/core.rs) it can lead to two things:
* fetching missing `parents` when a header is getting processed (via the `header_waiter`)
* fetching missing `payload` when a header is getting processed (via the `header_waiter`) 

This is an overlap of functionality with `block_synchronizer` and currently there is no way for each component to know whether the other has a pending request for the same entity. That means, if we blindly put a fetched certificate (e.x via the `block_synchronizer`) for processing to `core`, this can lead to redundant - duplicate fetching of payloads or parents which is totally inefficient and totally negates the work that is done via the `block_synchronizer`.

### Goal & Proposed Solution
The goal is to:
* remain consistent with the existing approach in processing & storing certificates & headers
* re-use as much as possible existing implementation

For that reason, regarding the `read_causal` is proposed to do the following:
1. Fetch the missing certificates via the `block_synchronizer`
2. Trigger the fetching of payloads via the `block_synchronizer`
3. _Causally_ send the fetched "missing" certificates to `core` one by one for processing.  When saying _causally_ , I mean send the certificates in order that will allow each certificate under processing to have all its parents available - so we avoid triggering any redundant re-fetching of parent certificates (see point A.2)

The above steps will make sure that we don't replicate checks and verifications to other components and that we also remain consistent to the existing approach. So we'll basically fulfil all the points of (A) & (B).

There is one piece though that is missing to ensure that we won't end up doing (potential) double processing when it comes to fetching the `payloads`. We should introduce a layer that will ensure that `header_waiter` and the `block_synchronizer` will not trigger overlapping requests to sync batches (some memoization) - and more specifically to ensure that the `header_waiter` will not trigger a sync while the `block_synchronizer` has already done. By introducing this layer we'll ensure that the step (3) above will not lead to fetching payloads again when we already have triggered that as part of step (2).

**Comment by akichidis** [comment]
After iterating on the above an approach to minimise the intervention in the existing components has been chosen. We decided to not perform any causal completion manually but rather let the core component perform that. Since on the `read_causal` request a single collection id is provided, we only want to fetch - if missing - this certificate from other peers. Once retrieved it is submitted to `core` for processing and causal completion (which already performs)  and then wait until the certificate appears in store to continue.

The [block synchronizer handler](https://github.com/MystenLabs/narwhal/pull/227) introduced to perform the missing certificate fetching from peers and also wait until the causal history has been completed. Once this is finished, then we are ready to continue and perform and perform the read_causal on the DAG

**Comment by huitseeker** [comment]
@akichidis Could you please review and make the issue content refer to the leftover tasks, if any?

**Comment by akichidis** [comment]
Given the approached we ended up following there is not current issue that is directly related to this so we are safe to close the ticket.
